### ISSUE TITLE
theorem's target anchor is sometimes placed in the heading of the previous page if heading contains a \vbox

### ISSUE THREAD

**Issue by boritomi** [latex3/hyperref#332]
Test case (pdflatex):
```
\documentclass{article}
\usepackage{fancyhdr}
\pagestyle{fancy}
\usepackage{hyperref}
\newtheorem{thm}{Theorem}
\textheight60pt
\begin{document}
1\\2\\3\\4\\5
\begin{thm}\label{th}
Theorem text.
\end{thm}
Ref to Theorem \ref{th}.
\end{document}
```

Checked with recent versions, each fails since 7.00w, 7.00v (and previous versions) are OK. (Perhaps since Feb 13, 2023 [new thm implementation]?)

**Comment by u-fischer** [comment]
I hate theorem anchors ;-). Every time you fix one instance another one breaks. 

I don't think that one can really resolve all the problems through patches in hyperref. I will try to trigger a change in the LaTeX commands instead, so that the anchor is directly where it belongs and hyperref can stop to mess around.

~~~~
\documentclass{article}
\makeatletter\let\@kernel@refstepcounter\refstepcounter\makeatother
\usepackage{fancyhdr}
\pagestyle{fancy}
\usepackage{hyperref}
\fancyhead{}
\makeatletter
\def\@thm#1#2{%
  \@kernel@refstepcounter{#1}%
  \@ifnextchar[{\@ythm{#1}{#2}}{\@xthm{#1}{#2}}}
\def\@begintheorem#1#2{\trivlist
   \item[\MakeLinkTarget{\@currentcounter}\hskip \labelsep{\bfseries #1\ #2}]\itshape}
\def\@opargbegintheorem#1#2#3{\trivlist
   \item[\MakeLinkTarget{\@currentcounter}\hskip \labelsep{\bfseries #1\ #2\ (#3)}]\itshape} 
          
\newtheorem{thm}{Theorem}
\textheight60pt
\begin{document}
1\\2\\3\\4\\5\\6

\begin{thm}\label{th}
Theorem text.
\end{thm}
Ref to Theorem \ref{th}.

\end{document}
~~~~

**Comment by u-fischer** [comment]
> inserting `\phantomsection` automatically for the theorems:

That is basically what my code above is doing: ` \item[\MakeLinkTarget{\@currentcounter}\hskip \labelsep{\bfseries #1\ #2\ (#3)}]\itshape}` (and imho the target should be at the begin of the label and not at the end.)  

But `\MakeLinkTarget` is better as it works also without hyperref. 

> In the long run, that would be very nice if hyperref could deal with theorems' target anchors better...

No in the long run it would be better if hyperref wouldn't have to deal with that at all.  It is simply a pain to have to patch these commands from the outside. It would be much better if amsthm etc would insert the targets directly. 

I have now opened a pull request to change at least the kernel definitions. https://github.com/latex3/latex2e/pull/1301.

**Comment by boritomi** [comment]
> No in the long run it would be better if hyperref wouldn't have to deal with that at all. It is simply a pain to have to patch these commands from the outside. It would be much better if amsthm etc would insert the targets directly.
> 
> I have now opened a pull request to change at least the kernel definitions. [latex3/latex2e#1301](https://github.com/latex3/latex2e/pull/1301).

I have to agree, however, it seems that your patch will break `amsthm` and maybe other libraries redefining `\@begintheorem` (or maybe also `\@thm`, `\@opargbegintheorem`, etc.) (try your code with `\usepackage{amsthm}` in the preamble).

**Comment by u-fischer** [comment]
> I have to agree, however, it seems that your patch will break amsthm

No, the kernel change should not affect them as they overwrite the definitions later (I checked that ...). So they only loose the target again.

**Comment by u-fischer** [comment]
with a current latex-dev (released a few day ago) this now works correctly.
